Ux 588 list box browser (#213)
acl-services/paprika: packages/ListBoxBrowser/README.md

# ListBoxBrowser

> It is a React component that allows users to select one or multiple options using two columns, similar to the experience of browsing a file manager in any giving Operative System.

Reviewer: "any given operating system"

> It is a React component that allows users to select one or multiple options using two columns, similar to the experience of browsing a file manager in any giving Operative System.

Reviewer 2: `any given operating system`

> It is a React component that allows users to select one or multiple options using two columns, similar to the experience of browsing a file manager in any giving Operative System.

Reviewer 2: Don't think we need to start off the sentence with "It is a React component".

Author: feel free to modify it :)

> It is a React component that allows users to select one or multiple options using two columns, similar to the experience of browsing a file manager in any giving Operative System.

Reviewer 2: ```suggestion
A two column layout that allows users to select one or multiple options. It is similar to the experience of browsing a file manager in any operating system.
```

## Installation

```sh
$ npm install @paprika/listbox-browser
```

or

```sh
$ yarn add @paprika/listbox-browser
```

## Props

```
browserTitle
children
data
defaultSelectedOptions
defaultSelectedView
hasBreadcrumb
hasError
height
isMulti
isParentSelectable
onChange
rootTitle
```

## Usage

> The `<ListBoxBrowser />` is primarily an uncontrolled component which received a `data` prop allowing the consumer to set the initial state.

Reviewer 2: ```suggestion
The `<ListBoxBrowser />` is primarily an uncontrolled component which receives a `data` prop which allows the consumer to set the initial state.
```

> Also, has two functions to set defaults for selected options and a default selected view for the browser column.

Reviewer 2: ```suggestion
It has two functions to set defaults. One for selected options and another for a default selected view for the browser column.
```

The `onChange` prop receives a parameter with an array of the selected options.

### Data prop

The most important prop for the `<ListBoxBrowser />` is the `data` prop, which initializes the state for the component.

> The `data` prop has a shape of an array of objects where each object requires to have at least a `label` attribute on it, and at least one of those items should include an `options` property.

Reviewer 2: ```suggestion
The `data` prop is an array of objects where each object requires to have at least a `label` attribute on it and at least one of those items should include an `options` property.
```

### Data prop shape

```js
{
	label: "require attribute",
	options: [{}{}{}], // more object with the same shape
}
```

The **object** can have an options attribute, which is an array of the same kind of objects.

A data prop with **options** looks like:

```js
const data = [
   {
        label: "NBA",
        options: [{ label: "Toronto Raptors"}, {...}]
   },
   {
        label: "NHL",
        options: [{ label: "Vancouver Canucks"}, {...}]
   },
   {
        label: "MLS",
        options: [{ label: "Vancouver Whitecaps"}, {...}]
   },
]
```

> Recursively the component runs over all the options on the array and assembles the appropriated UI for the user.

Reviewer 2: ```suggestion
Recursively the component runs over all the options on the array and assembles the UI for the user.
```

> **NOTE** You can add extra properties to the objects; those are and pass to you back on the `onChange` function or when using `defaultSelectedOptions`, `defaultSelectedView` functions.

Reviewer 2: This one i'm not really sure about the grammar on this part
"those are and pass to you back on the `onChange` function or when using `defaultSelectedOptions`, `defaultSelectedView` functions.".

Maybe pro @alexzherdev can chime in with his english skills :)

---

#### Basic example

```js
import ListBoxBrowser from "@paprika/listbox-browser";

export default function App() {
   const data = [{...}...{...}]; // your data;
   return <ListBoxBrowser data={data} />
}
```

---

#### ListBoxBrowser.SelectedOptions

By default, the `<ListBoxBrowser />` does not display a list of the selected options.  
But you can activate this feature to display multiple options using the `<ListBoxBrowser.SelectedOptions />` subcomponent.

```js
<ListboxBrowser data={data}>
   <ListBoxBrowser.SelectedOptions />
</ListBoxBrowser>
```

#### defaultSelectedOptions

> The `defaultSelectedOptions` received a function to be executed to determine if an option should start as selected or not. This function is called as many times as options are pass down on the data prop.

Reviewer 2: ```suggestion
The `defaultSelectedOptions` receives a function to be executed to determine if an option should start as selected or not. This function is called as options are passed down on the data prop.
```

Ex.

```js
const data = [
	{ key: 1, label: "one", options: [{...}]},
	{ key: 2, label: "two", options: [{...}]},
	...
]

<ListBoxBrowser
	data={data}
	defaultSelectedView={(option) =>  (option.key === 1)}  />

/**
  The previous result on option with `key 1` to be selected */
```

#### defaultSelectedView

> This prop worked exactly like the previous one, receiving a function which you can use to compare and decide what option should be the one as initial view.

Reviewer 2: ```suggestion
This prop works exactly like the `defaultSelectedOptions`. It receives a function which you can use to compare and decide what option should be the one as initial view.
```

> **Note**: Be sure to selected an `option` property with `options`; otherwise, this function doesn't work and fall back to the default behaviour, which is to select the first option with options.

Reviewer 2: ```suggestion
**Note**: Be sure to select an `option` property with `options`. Otherwise, this function doesn't work and falls back to the default behaviour, which selects the first option with options.
```

### Lazy loading

> This component provides a way to load async options. To achieve this, you should declare the `options` array property empty `[]`, this indicates to the component to fire the `onFetch={option => ()}` prop when the user interacts with it.

Reviewer 2: ```suggestion
This component provides a way to load async options. To achieve this, you can declare the `options` array property empty `[]`. This indicates to the component to fire the `onFetch={option => ()}` prop when the user interacts with it.
```

#### Lazy loading example

To help you to sync your data with the newest options, the ListBoxBrowser provides you with some tools to make it easier:

A Subcomponent `<ListBoxBrowser.Browser isLoading />` which lets you set the right column into a pending state.

> And a `ListBoxBrowser.findOption(data, fn)` which helps you in finding an option in your data so you can modify it.

Reviewer 2: ```suggestion
A `ListBoxBrowser.findOption(data, fn)` which helps you in finding an option in your data so you can modify it.
```
